comms/src/blazingdb/transport/io/reader_writer.cpp

```diff
@@ -29,42 +29,58 @@ namespace io {
 // numBuffers should be equal to number of threads
 PinnedBufferProvider::PinnedBufferProvider(std::size_t sizeBuffers,
                                            std::size_t numBuffers) {
+  this->bufferSize = sizeBuffers;
   this->buffer_counter = numBuffers;
-  for (int bufferIndex = 0; bufferIndex < numBuffers; bufferIndex++) {
+  this->allocations = std::vector<char *>(1);
+  cudaError_t err = cudaMallocHost((void **)&allocations[0], numBuffers * sizeBuffers * 5);
+  if (err != cudaSuccess) {
+    throw std::exception();
+  }
+  for (int bufferIndex = 0; bufferIndex < numBuffers * 5; bufferIndex++) {
     PinnedBuffer *buffer = new PinnedBuffer();
-    buffer->size = sizeBuffers;
-    this->bufferSize = sizeBuffers;
-    cudaError_t err = cudaMallocHost((void **)&buffer->data, sizeBuffers);
-    if (err != cudaSuccess) {
-      throw std::exception();
-    }
+    buffer->size = this->bufferSize;
+    buffer->data = allocations[0] + bufferIndex * this->bufferSize;
     this->buffers.push(buffer);
   }
+  this->buffer_counter =  numBuffers * 5;
 }
 
 PinnedBufferProvider::~PinnedBufferProvider(){
-    getPinnedBufferProvider().freeAll();
+  getPinnedBufferProvider().freeAll();
 }
 
 // TODO: consider adding some kind of priority
 // based on when the request was made
 PinnedBuffer *PinnedBufferProvider::getBuffer() {
   std::unique_lock<std::mutex> lock(inUseMutex);
   if (this->buffers.empty()) {
-    // cv.wait(lock, [this] { return !this->buffers.empty(); });
-    // if wait fail use:
-    this->grow();
+    //if (this->buffer_counter >= 100){
+    //    cv.wait(lock, [this] { return !this->buffers.empty(); });        
+    //}else{
+        this->grow();
+    //}
+    
   }
   PinnedBuffer *temp = this->buffers.top();
   this->buffers.pop();
   return temp;
 }
 
 void PinnedBufferProvider::grow() {
-  this->buffer_counter++;
+  std::unique_lock<std::mutex> lock(inUseMutex);
+    
+
   PinnedBuffer *buffer = new PinnedBuffer();
   buffer->size = this->bufferSize;
-  cudaError_t err = cudaMallocHost((void **)&buffer->data, this->bufferSize);
+  allocations.resize(allocations.size() + 1);
+  cudaError_t err = cudaMallocHost((void **)&allocations[allocations.size() -1], this->bufferSize * 100);
```

Comment:
I would suggest growing the buffer pool by `numBuffers/2`  and have `numBuffers` be stored as a member of PinnedBufferProvider. That way we dont have another random hardcoded number. 
I suggest `numBuffers/2` because that is a common pattern in memory pools to grow by 50% of the initial amount.

Reply:
in all honesty i am kind of opposed to merging this in until we add something that will alos shrink the pool